cpp/src/arrow/compare.cc

```diff
@@ -305,6 +305,11 @@ class RangeDataEqualsImpl {
   Status Visit(const StructType& type) {
     const int32_t num_fields = type.num_fields();
 
+    if (left_.child_data.size() != static_cast<size_t>(num_fields) ||
```

Comment:
Can we please separate out orthogonal changes?

Reply:
Will do.